src/compiler/transformers/es2015.ts

```diff
@@ -1089,19 +1100,43 @@ namespace ts {
                 insertCaptureThisForNodeIfNeeded(prologue, constructor);
             }
 
-            const block = factory.createBlock(
+            const bodyBlock = factory.createBlock(
                 setTextRange(
                     factory.createNodeArray(
-                        concatenate(prologue, statements)
+                        [
+                            ...existingPrologue,
+                            ...prologue,
+                            ...visitNodes(factory.createNodeArray(preSuperStatements), visitor, isStatement),
+                            ...statements
+                        ]
                     ),
                     /*location*/ constructor.body.statements
                 ),
                 /*multiLine*/ true
             );
 
-            setTextRange(block, constructor.body);
+            setTextRange(bodyBlock, constructor.body);
+            return bodyBlock;
+        }
 
-            return block;
+        function splitConstructorBodyStatementsOnSuper(originalBodyStatements: NodeArray<Statement>, existingPrologue: Statement[]) {
+            for (let i = existingPrologue.length; i < originalBodyStatements.length; i += 1) {
+                const superCall = getSuperCallFromStatement(originalBodyStatements[i]);
+                if (superCall) {
+                    // With a super() call, split the statements into pre-super() and 'body' (post-super())
+                    return {
+                        bodyStatements: factory.createNodeArray(originalBodyStatements.slice(i + 1)),

```

Comment:
The `visitNodes` function accepts a `start` and a `count`. Rather than allocating a new array here, you could just return the offsets here and pass the requisite offsets into `visitNodes` above.

Reply:
Great, done for `superStatementIndex` instead of `preSuperStatements`. I don't see how to get that working with `bodyStatements` though, since it's passed directly to `factory.copyStandardPrologue`?

Reply:
For `copyStandardPrologue` you can just use `.slice` since that takes a `readonly Statement[]` rather than a `NodeArray<Statement>`, so there's no need to go through the effort of converting it into a `NodeArray`. That said, we could possibly add an offset to `copyStandardPrologue` similar to the one we have for `copyCustomPrologue`.